src/modules/job-manager/plugins/dependency-after.c

```diff
@@ -0,0 +1,563 @@
+        }
+    }
+}
+
+/*
+ *  Raise exceptions for all unhandled depednencies in list `l`.
+ */
+static void raise_exceptions (flux_plugin_t *p, zlistx_t *l)
+{
+    if (l) {
+        struct after_info *after;
+        FOREACH_ZLISTX (l, after) {
+            if (flux_jobtap_raise_exception (p,
+                                             after->depid,
+                                             "dependency",
+                                             0,
+                                             "%s %s can never be satisfied",
+                                             "dependency",
+                                             after->description) < 0)
+                flux_log_error (flux_jobtap_get_flux (p),
+                                "id=%ju: unable to raise exception for %s",
+                                (uintmax_t) after->depid,
+                                after->description);
+        }
+        /*  N.B. = entry will be deleted at list destruction */
+    }
+}
+
+/*  If this job has any outstanding after-dependency references, then the
+ *   job has transitioned from dependency state directly to cleanup
+ *   (e.g. due to cancelation) and the refs must be cleaned up. This avoids
+ *   prerequisite jobs from emitting erroneous dependencies for completed
+ *   jobs.
+ */
+static void release_dependency_references (flux_plugin_t *p)
+{
+    flux_t *h = flux_jobtap_get_flux (p);
+    zlistx_t *l;
+    if ((l = after_refs_check (p))) {
+        struct after_ref *ref = zlistx_first (l);
+        while (ref) {
+            /*  For each after_ref entry, check to ensure the job and its
+             *   after dependencies list still exists. If so, remove this
+             *   job's entry from the list.
+             */
+            zlistx_t *after_list = after_list_check (p, ref->id);
+            if (after_list == ref->list) {
+                void *handle = zlistx_find (after_list, ref->info);
+                if (handle && zlistx_delete (ref->list, handle) < 0) {
+                    flux_log_error (h, "%s: %s: zlistx_delete",
+                                    "dependency-after",
+                                    "release_references");
+                }
+            }
+            ref = zlistx_next (l);
+        }
+    }
+
+    /*  Destroy this job's dependency reference list.
+     */
+    if (flux_jobtap_job_aux_delete (p, FLUX_JOBTAP_CURRENT_JOB, l) < 0)
+        flux_log_error (h, "release_references: flux_jobtap_job_aux_delete");
+}
+
+static int release_dependent_jobs (flux_plugin_t *p, zlistx_t *l)
+{
+    flux_t *h = flux_jobtap_get_flux (p);
+    flux_job_result_t result;
+
+    if (l == NULL)
+        return 0;
+
+    if (flux_jobtap_get_job_result (p,
+                                    FLUX_JOBTAP_CURRENT_JOB,
+                                    &result) < 0) {
+        flux_log_error (h, "dependency-after: flux_jobtap_get_result");
+        return -1;
+    }
```

Comment:
Not a comment on this code, but in `flux_jobtap_get_job_result()`, errno is not set on `json_unpack_ex()` error, but it appears from this usage and from other error paths that it should be.

Reply:
Yes, should be. thanks.